File: 623-add-one-row-to-tree/623-add-one-row-to-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        dummy,dummy.left=TreeNode(None),root
        row=[dummy]
        for _ in range(depth-1):
            row=[child for node in row for child in [node.left,node.right] if child]
        for node in row:
            node.left,node.left.left=TreeNode(val),node.left
            node.right,node.right.right=TreeNode(val),node.right
        return dummy.left

File: 623-add-one-row-to-tree/623-add-one-row-to-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* addOneRow(TreeNode* root, int val, int depth) {
        if(depth==1){
            TreeNode* newNode=new TreeNode(val);
            newNode->left=root;
            return newNode;
        }
        queue<TreeNode*>q;
        q.push(root);
        //BFS
        for(int i=1;i<depth-1;i++){
            int n=q.size();
            for(int i=0;i<n;i++){
                TreeNode* top=q.front();
                q.pop();
                if(top->left)q.push(top->left);
                if(top->right)q.push(top->right);
            }
        }
        while(!q.empty()){
            TreeNode* temp=q.front()->left;
            q.front()->left=new TreeNode(val);
            q.front()->left->left=temp;
            temp=q.front()->right;
            q.front()->right=new TreeNode(val);
            q.front()->right->right=temp;
            q.pop();
        }
        return root;
    }
};